src/domain/mongoose_domain_loader.erl

```diff
@@ -18,81 +60,114 @@ load_data_from_base(FromId, PageSize) ->
-        [] -> %% Skipping this time
-            ?LOG_WARNING(#{what => domain_check_for_updates_skipped,
-                           from_id => FromId}),
-            FromId;
-        Rows ->
-            Now = mongoose_domain_gaps:new_timestamp(),
-            case mongoose_domain_gaps:check_for_gaps(Rows, Now) of
-                ok ->
-                    case check_if_id_is_still_relevant(FromId, Rows) of
-                        [] -> FromId;
-                        RowsToApply ->
-                            PageMaxId = row_to_id(lists:last(RowsToApply)),
-                            apply_changes(RowsToApply),
-                            check_for_updates(PageMaxId, PageSize)
-                    end;
-                restart ->
-                    service_domain_db:restart(),
-                    FromId;
-                wait ->
-                    FromId
-            end
-    catch Class:Reason:StackTrace ->
-        %% Don't allow to crash in the critical code,
-        %% once we've started.
-        ?LOG_ERROR(#{what => domain_check_for_updates_failed,
-                     from_id => FromId,
-                     class => Class, reason => Reason,
-                     stacktrace => StackTrace}),
-        FromId
-    end.
+    remove_domains(OutdatedDomains),
+    {ok, #{count => length(OutdatedDomains)}}.
 
-check_if_id_is_still_relevant(FromId, Rows) ->
-    MinId = row_to_id(hd(Rows)),
-    %% FromID should be always equal or less than MinID,
-    %% see check_for_updates/2 function for more details
-    if
-        FromId =:= MinId ->
-            tl(Rows);
-        FromId =:= MinId - 1 ->
-            %% looks like someone completely erased the events table
-            %% this should not happen, but we are still fine.
-            Rows;
-        FromId < MinId - 1 ->
-            %% Looks like this node has no DB connection for a long time.
-            %% But the event log in the DB has been truncated by some other node
-            %% meanwhile. We have to load the whole set of data from DB.
-            Text = <<"DB domain log had some updates to domains deleted,"
-                     " which we have not applied yet. Have to crash.">>,
-            ?LOG_CRITICAL(#{what => events_log_out_of_sync,
-                            text => Text, min_db => MinId, from_id => FromId}),
-            service_domain_db:restart(),
-            []
-    end.
+check_for_updates() ->
+    MinMax = mongoose_domain_sql:get_minmax_event_id(),
+    State = mongoose_loader_state:get(undefined),
+    check_for_updates(MinMax, State).
+
+check_for_updates({null, null}, State) ->
+    State;
+check_for_updates({Min, Max}, #{min_event_id := Min, max_event_id := Max} = State) ->
+    State;
+check_for_updates(MinMax = {Min, Max}, State) when is_integer(Min), is_integer(Max) ->
+    #{gaps := OldGaps, min_event_id := OldMin, max_event_id := OldMax} =
+        State,
+    {MinEventId, MaxEventId} = limit_max_id(OldMax, MinMax, 1000),
```

Comment:
It looks to me like we only handle the single page without moving to the next one.

Reply:
added looping here